compiler/optimizer/abstractinterpreter/InliningProposal.cpp

```diff
@@ -0,0 +1,260 @@
+
+   TR_ASSERT_FATAL(_idt, "Must have an IDT");
+
+   char header[1024];
+   sprintf(header,"#Proposal: %d methods inlined into %s, cost: %d", numMethodsToInline, _idt->getRoot()->getName(comp->trMemory()), getCost());
+
+   if (traceBIProposal)
+      traceMsg(comp, "%s\n", header);
+   if (verboseInlining)
+      TR_VerboseLog::writeLineLocked(TR_Vlog_BI, header);
+
+   TR::deque<TR::IDTNode*, TR::Region&> idtNodeQueue(comp->trMemory()->currentStackRegion());
+   idtNodeQueue.push_back(_idt->getRoot());
+
+   //BFS
+   while (!idtNodeQueue.empty())
+      {
+      TR::IDTNode* currentNode = idtNodeQueue.front();
+      idtNodeQueue.pop_front();
+      const int32_t index = currentNode->getGlobalIndex();
+
+      if (index != -1) //do not print the root node
+         {
+         char line[1024];
+         sprintf(line, "#Proposal: #%d : #%d %s @%d -> bcsz=%d %s target %s, benefit = %f, cost = %d, budget = %d",
+            currentNode->getGlobalIndex(),
+            currentNode->getParentGlobalIndex(),
+            _nodes->isSet(index + 1) ? "INLINED" : "NOT inlined",
+            currentNode->getByteCodeIndex(),
+            currentNode->getByteCodeSize(),
+            currentNode->getResolvedMethodSymbol()->signature(comp->trMemory()),
+            currentNode->getName(comp->trMemory()),
+            currentNode->getBenefit(),
+            currentNode->getCost(),
+            currentNode->getBudget()
+         );
+
+         if (traceBIProposal)
+            traceMsg(comp, "%s\n",line);
+         if (verboseInlining)
+            TR_VerboseLog::writeLineLocked(TR_Vlog_BI, line);
+         }
+
+      //process children
+      const uint32_t numChildren = currentNode->getNumChildren();
+
+      for (uint32_t i = 0; i < numChildren; i++)
+         idtNodeQueue.push_back(currentNode->getChild(i));
+
+      }
+
+   traceMsg(comp, "\n");
+   }
+
+
+void TR::InliningProposal::addNode(TR::IDTNode *node)
+   {
+   ensureBitVectorInitialized();
+
+   const int32_t index = node->getGlobalIndex() + 1;
+   if (_nodes->isSet(index))
+      {
+      return;
+      }
+
+   _nodes->set(index);
+
+   _cost = 0;
+   _benefit = 0;
+   }
+
+bool TR::InliningProposal::isEmpty()
+   {
+   if (!_nodes)
+      return true;
+
+   return _nodes->isEmpty();
+   }
+
+uint32_t TR::InliningProposal::getCost()
+   {
+   if (_cost == 0)
+      {
+      computeCostAndBenefit();
+      }
+
+   return _cost;
+   }
+
+double TR::InliningProposal::getBenefit()
+   {
+   if (_benefit == 0)
+      {
+      computeCostAndBenefit();
+      }
+
+   return _benefit;
+   }
+
+void TR::InliningProposal::computeCostAndBenefit()
+   {
+   _cost = 0;
+   _benefit = 0;
+
+   if (!_idt)
+      return;
+
+   TR_BitVectorIterator bvi(*_nodes);
+   int32_t idtNodeIndex;
+
+   while (bvi.hasMoreElements())
+      {
+      idtNodeIndex = bvi.getNextElement();
+      IDTNode *node = _idt->getNodeByGlobalIndex(idtNodeIndex - 1);
+      if (node == NULL)
+         {
+         continue;
```

Comment:
I have checked code and run some testcases. It should not be `NULL` in any situation. I checked our git history and this was introduced at the beginning of the project, so I assume the original contributors just added it in case that the `node` is `NULL`, or they have amended the structure of IDT later but forgot to change this part.